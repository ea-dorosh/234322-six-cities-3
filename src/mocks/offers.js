import {getRandomNumber, getRandomElement} from "../utils";

const PRICE = {
  MIN: 50,
  MAX: 500
};

const offersNames = [
  `Beautiful & luxurious apartment at great location`,
  `Nice, cozy, wood and stone place`,
  `Wood and stone place`,
  `Nice, cozy, warm big bed apartment`
];

const holderNames = [
  `Angelina`,
  `Max`,
];

const offersTypes = [
  `apartment`,
  `room`,
  `house`,
  `hotel`
];

const offersCords = [
  [52.3909553943508, 4.85309666406198],
  [52.369553943508, 4.85309666406198],
  [52.3909553943508, 4.929309666406198],
  [52.3809553943508, 4.939309666406198]
];

const createOffer = (index) => {
  const bedRoomQuantity = getRandomNumber(1, 6);
  const maxGuestQuantity = bedRoomQuantity * 2;
  const holderName = getRandomElement(holderNames);

  return {
    img: `img/apartment-0${getRandomNumber(1, 3)}.jpg`,
    photos: [
      `img/apartment-0${getRandomNumber(1, 3)}.jpg`,
      `img/apartment-0${getRandomNumber(1, 3)}.jpg`,
      `img/apartment-0${getRandomNumber(1, 3)}.jpg`,
      `img/apartment-0${getRandomNumber(1, 3)}.jpg`,
      `img/apartment-0${getRandomNumber(1, 3)}.jpg`,
      `img/apartment-0${getRandomNumber(1, 3)}.jpg`,
    ],
    isPremium: !!(index % 2),
    price: getRandomNumber(PRICE.MIN, PRICE.MAX),
    name: offersNames[index],
    type: getRandomElement(offersTypes),
    rating: parseFloat(`${getRandomNumber(2, 4)}.${getRandomNumber(0, 9)}`),
    description: `An independent House, strategically located between Rembrand Square and National Opera, but where the bustle of the city comes to rest in this alley flowery and colorful.`,
    bedRoomQuantity,
    maxGuestQuantity,
    details: [
      `Wi-Fi`,
      `Washing machine`,
      `Towels`,
      `Heating`,
      `Coffee machine`,
      `Baby seat`,
      `Kitchen`,
      `Dishwasher`,
      `Cabel TV`,
      `Fridge`,
    ],
    holder: {
      holderName,
      img: `img/avatar-${holderName.toLowerCase()}.jpg`,
      isSuper: !(index % 2),
    },
    coords: {
      x: offersCords[index][0],
      y: offersCords[index][1],
    }
  };
};

const createOffers = (amount) => {
  const offers = [];

  for (let i = 0; i < amount; i++) {
    const offer = createOffer(i);

    offers.push(offer);
  }
  return offers;

};


export {createOffers};
