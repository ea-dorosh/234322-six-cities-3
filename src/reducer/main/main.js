import {extend} from '../../utils.js';


const initialState = {
  sortType: `Popular`,
  marker: null,
};


const ActionType = {
  SORT_OFFERS: `SORT_OFFERS`,
  HIGHLIGHT_MARKER: `HIGHLIGHT_MARKER`,
};

const ActionCreator = {
  sortOffers: (sortType) => ({
    type: ActionType.SORT_OFFERS,
    payload: sortType,
  }),
  highlightMarker: (offer) => ({
    type: ActionType.HIGHLIGHT_MARKER,
    payload: offer,
  }),
};

const reducer = (state = initialState, action) => {
  switch (action.type) {

    case ActionType.SORT_OFFERS:
      return extend(state, {sortType: action.payload});

    case ActionType.HIGHLIGHT_MARKER:
      return extend(state, {marker: action.payload});
  }
  return state;
};

export {reducer, ActionType, ActionCreator};
