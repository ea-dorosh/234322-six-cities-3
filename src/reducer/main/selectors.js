import NameSpace from "../name-space.js";


const NAME_SPACE = NameSpace.MAIN;

export const getMarker = (state) => {
  return state[NAME_SPACE].marker;
};

export const getSortType = (state) => {
  return state[NAME_SPACE].sortType;
};
