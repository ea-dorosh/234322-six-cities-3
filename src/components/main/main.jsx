import React from "react";
import PropTypes from "prop-types";
import PlaceCardList from "../place-card-list/place-card-list.jsx";
import Map from "../map/map.jsx";
import LocationsList from "../locations-list/location-list.jsx";
import SortOptions from "../sort-options/sort-options.jsx";
import {connect} from "react-redux";
import {CITIES} from "../../utils.js";
import {ActionCreator} from "../../reducer/main/main.js";
import MainEmpty from "../main-empty/main-emty.jsx";
import {getMarker} from "../../reducer/main/selectors.js";
import {selectOffers, getActiveCity} from "../../reducer/offers/selectors.js";
import {getAuthorizationStatus} from "../../reducer/user/selectors";
import {Operation as UserOperation} from "../../reducer/user/user";
import {AuthorizationStatus} from "../../reducer/user/user.js";
import {Link} from 'react-router-dom';


const Main = (props) => {
  const {
    // eslint-disable-next-line react/prop-types
    authorizationStatus,
    activeCity,
    offers,
    // eslint-disable-next-line react/prop-types
    marker, handleOfferHover,
  } = props;

  const advertsCount = offers.length;

  return (
    <div className="page page--gray page--main">
      <header className="header">
        <div className="container">
          <div className="header__wrapper">
            <div className="header__left">
              <a className="header__logo-link header__logo-link--active">
                <img className="header__logo" src="img/logo.svg" alt="6 cities logo" width="81" height="41"/>
              </a>
            </div>
            <nav className="header__nav">
              <ul className="header__nav-list">
                <li className="header__nav-item user">
                  <Link
                    to={`/dev-auth`}
                    className="header__nav-link header__nav-link--profile" href="#">
                    <div className="header__avatar-wrapper user__avatar-wrapper">
                    </div>
                    <span className="header__user-name user__name">
                      {authorizationStatus === AuthorizationStatus.NO_AUTH ? `Sign in` : `Oliver.conner@gmail.com`}
                    </span>
                  </Link>
                </li>
              </ul>
            </nav>
          </div>
        </div>
      </header>

      <main className={`page__main page__main--index ${ advertsCount ? `` : `page__main--index-empty` }`}>
        <h1 className="visually-hidden">Cities</h1>
        <div className="tabs">
          <section className="locations container">
            <LocationsList/>
          </section>
        </div>
        <div className="cities">
          { advertsCount ?
            <div className="cities__places-container container">
              <section className="cities__places places">
                <h2 className="visually-hidden">Places</h2>
                <b className="places__found">{advertsCount} places to stay in {activeCity.name}</b>
                <SortOptions/>
                <PlaceCardList
                  offers={offers}
                  handleOfferHover={handleOfferHover}
                  listClass={CITIES}
                />
              </section>
              <div className="cities__right-section">
                <section className="cities__map map">
                  <Map
                    offers={offers}
                    marker={marker}
                    activeCity={activeCity}
                  />
                </section>
              </div>
            </div>
            :
            <MainEmpty
              activeCity={activeCity}
            />
          }
        </div>
      </main>
    </div>
  );
};

Main.propTypes = {
  offers: PropTypes.array,
  activeCity: PropTypes.shape({
    location: PropTypes.shape({
      latitude: PropTypes.number,
      longitude: PropTypes.number,
      zoom: PropTypes.number
    }),
    name: PropTypes.string,
  }).isRequired,
  handleOffersSort: PropTypes.func,
  sortType: PropTypes.string,
};


const mapStateToProps = (state) => {
  return {
    authorizationStatus: getAuthorizationStatus(state),
    activeCity: getActiveCity(state),
    offers: selectOffers(state),
    marker: getMarker(state)
  };
};

const mapDispatchToProps = (dispatch) => ({
  login(authData) {
    dispatch(UserOperation.login(authData));
  },

  handleOfferHover(offer) {
    dispatch(ActionCreator.highlightMarker(offer));
  }
});


export default connect(mapStateToProps, mapDispatchToProps)(Main);

