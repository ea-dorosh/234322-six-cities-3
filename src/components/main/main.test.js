import React from "react";
import renderer from "react-test-renderer";
import {Main} from "./main.jsx";
import {citiesMock, offersMock} from "../../mocks/offers_for_test";

it(`Should Main render correctly`, () => {
  const tree = renderer
    .create(
        <Main
          activeCity={citiesMock[0]}
          offers={offersMock}
          marker={offersMock[0]}
          handleOfferHover={()=>{}}
        />
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
