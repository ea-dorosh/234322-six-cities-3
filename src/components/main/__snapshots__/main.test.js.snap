// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should Main render correctly 1`] = `
<div
  style={
    Object {
      "fontSize": "60px",
    }
  }
>
  <p>
    Connection...
  </p>
</div>
`;
